runtimes/pythonrt/runner_manager_remote_client.go

```diff
@@ -19,9 +27,111 @@ type RemoteRuntimeConfig struct {
-	logger         *zap.Logger
-	remoteManagers []runner_managerv1connect.RunnerManagerServiceClient
+type managedRunnerManagerClient struct {
+	client runner_managerv1connect.RunnerManagerServiceClient
+	addr   string
+}
+
+func newManagedRunnerManager(addr string) (*managedRunnerManagerClient, error) {
+	runner := runner_managerv1connect.NewRunnerManagerServiceClient(http.DefaultClient, addr, connect.WithGRPC())
+
+	resp, err := runner.Health(context.Background(), connect.NewRequest(&rmv1.HealthRequest{}))
+	if err != nil {
+		return nil, fmt.Errorf("could not verify runner manager health")
+	}
+
+	if resp.Msg.Error != "" {
+		return nil, fmt.Errorf("runner manager health: %w", err)
+	}
+
+	return &managedRunnerManagerClient{
+		client: runner,
+		addr:   addr,
+	}, nil
+}
+
+type remoteRunnersManager struct {
+	managedRunnerManagers []*managedRunnerManagerClient
+	managersLock          *sync.Mutex
+	nextClientIndex       int
+	activeUseCount        map[int]int
+	maxUsePerClient       int
+	logger                *zap.Logger
+
+	runnersLock             *sync.Mutex
+	runnerIDToRemoteManager map[string]*managedRunnerManagerClient //map runnerID to runner manager
+	runnerIDToClient        map[string]*RunnerClient               //map runnerID to RunnerClient
+
+	cfg RemoteRuntimeConfig
+}
+
+func newRemoteRunnerManager(l *zap.Logger, cfg RemoteRuntimeConfig) *remoteRunnersManager {
+	return &remoteRunnersManager{
+		managedRunnerManagers:   make([]*managedRunnerManagerClient, 0),
+		managersLock:            &sync.Mutex{},
+		runnersLock:             &sync.Mutex{},
+		runnerIDToRemoteManager: make(map[string]*managedRunnerManagerClient),
+		runnerIDToClient:        make(map[string]*RunnerClient),
+		nextClientIndex:         0,
+		activeUseCount:          make(map[int]int),
+		maxUsePerClient:         50, //TODO: get this from the client, need to extened rpc
+		logger:                  l,
+		cfg:                     cfg,
+	}
+}
+
+func (p *remoteRunnersManager) addRemoteManager(cli *managedRunnerManagerClient) {
+	p.managersLock.Lock()
+	defer p.managersLock.Unlock()
+	p.managedRunnerManagers = append(p.managedRunnerManagers, cli)
+}
+
+// func (p *remoteRunnersManager) removeRemoteManager(cli *managedRunnerManagerClient) {
+// 	p.managersLock.Lock()
+// 	defer p.managersLock.Unlock()
+// 	for i, c := range p.managedRunnerManagers {
+// 		if c == cli {
+// 			p.managedRunnerManagers = append(p.managedRunnerManagers[:i], p.managedRunnerManagers[i+1:]...)
+// 			return
+// 		}
+// 	}
+// }
+
+func (p *remoteRunnersManager) acquireRemoteManager() (*managedRunnerManagerClient, error) {
+	p.managersLock.Lock()
+	defer p.managersLock.Unlock()
+	if len(p.managedRunnerManagers) == 0 {
+		return nil, errors.New("no clients available")
+	}
+
+	// Algorithm, find the client with the least active use count
+	var cli *managedRunnerManagerClient
+	minUsage := p.maxUsePerClient
+
+	for i, c := range p.managedRunnerManagers {
+		if p.activeUseCount[i] < minUsage {
```

Comment:
Instead of aligned lists, have a data structure for client + count.
Also have a data structure for `managedRunnerManagers` with the actions you want that does the locking and more.